src/autora/state/wrapper.py

```diff
@@ -85,3 +88,164 @@ def experimentalist(params):
+            to the state fields
+        output_field_name: A string that maps the return value of f to the state field
+
+    Examples:
+        >>> from autora.state.delta import State, Delta
+        >>> from dataclasses import dataclass, field
+        >>> import pandas as pd
+        >>> from typing import List, Optional
+
+        The `State` it operates on needs to have the metadata described in the state module:
+        >>> @dataclass(frozen=True)
+        ... class S(State):
+        ...     conditions: List[int] = field(metadata={"delta": "replace"})
+
+        We define an arbitraty function
+        >>> def experimentalist(x):
+        ...     new_conditions = [c + 10 for c in x]
+        ...     return new_conditions
+
+        We can now use the wrapper to get back a function that works on a state. Here, we need to
+        map the input arguments and the return value of the function to the fields of the state:
+        >>> experimentalist_on_state = wrap_to_use_state_general(
+        ...                                                 f=experimentalist,
+        ...                                                 input_state_mapping={'x':'conditions'},
+        ...                                                 output_field_name='conditions')
+
+        >>> experimentalist_on_state(S(conditions=[1,2,3,4]))
+        S(conditions=[11, 12, 13, 14])
+
+        >>> experimentalist_on_state(S(conditions=[101,102,103,104]))
+        S(conditions=[111, 112, 113, 114])
+
+        >>> from autora.variable import VariableCollection, Variable
+        >>> from sklearn.base import BaseEstimator
+        >>> from sklearn.linear_model import LinearRegression
+
+        We can use other states with different fields:
+        >>> @dataclass(frozen=True)
+        ... class T(State):
+        ...     variables: VariableCollection  # field(metadata={"delta":... }) omitted ∴ immutable
+        ...     experiment_data: pd.DataFrame = field(metadata={"delta": "extend"})
+        ...     model: Optional[BaseEstimator] = field(metadata={"delta": "replace"}, default=None)
+
+        >>> t = T(
+        ...     variables=VariableCollection(independent_variables=[Variable("x")],
+        ...                                  dependent_variables=[Variable("y")]),
+        ...     experiment_data=pd.DataFrame({"x": [0,1,2,3,4], "y": [2,3,4,5,6]})
+        ... )
+
+        And define other functions:
+        >>> def theorist(x: pd.DataFrame, v: VariableCollection, **kwargs):
+        ...     ivs = [v.name for v in v.independent_variables]
+        ...     dvs = [v.name for v in v.dependent_variables]
+        ...     X, y = x[ivs], x[dvs]
+        ...     new_model = LinearRegression(fit_intercept=True).set_params(**kwargs).fit(X, y)
+        ...     return new_model
+
+        Again, we use the wrapper:
+        >>> theorist_on_state = wrap_to_use_state_general(
+        ...                                         theorist,
```

Comment:
do we want to include keywords?

```suggestion
        ...                                         f=theorist,
```